server/pulp/server/lazy/streamer.py

```diff
@@ -0,0 +1,212 @@
+            * The requested URL is checked to ensure it has a valid signature
+              from the Pulp server.
+            * The unit specified by the request is looked up in the Pulp unit
+              catalog and the correct downloader is retrieved based on the specific
+              file requested.
+            * The file is downloaded using the Nectar downloader and the content
+              is streamed to the client as it is received.
+
+        :param request: the request to process.
+        :type  request: twisted.web.server.Request
+        """
+        reactor.callInThread(self._download, request)
+        return NOT_DONE_YET
+
+    def _download(self, request):
+        """
+        Download the requested content using the content unit catalog. This method also
+        dispatches a celery task that causes Pulp to download the newly cached unit.
+
+        :param request: The content request.
+        :type  request: twisted.web.server.Request
+        """
+        # The request.uri might contain GET params, which we don't want.
+        path = urlparse(request.uri).path
+        relative_path = path.lstrip('/')
+
+        with Responder(request) as responder:
+            try:
+                # Collect the plugin and instance configuration to use for the downloader
+                catalog_entry = model.LazyCatalogEntry.objects(relative_path=relative_path).get()
+                try:
+                    importer, config = repo_controller.get_importer_by_id(catalog_entry.plugin_id)
+                except PluginNotFound:
+                    logger.error(_('Catalog entry for {rel} references a plugin id, {id}, '
+                                   'which is not a valid id.'.format(rel=relative_path,
+                                                                     id=catalog_entry.plugin_id)))
+                    request.setResponseCode(INTERNAL_SERVER_ERROR)
+                    return
+
+                downloader = importer.get_downloader(config, catalog_entry.url,
+                                                     **catalog_entry.data)
+                downloader.event_listener = StreamerNectarListener(request)
+                download_request = nectar_request.DownloadRequest(catalog_entry.url, responder)
+                downloader.download_one(download_request, events=True)
+
+                # Dispatch the Pulp task to download the unit.
+                content_controller.queue_download(catalog_entry.plugin_id,
+                                                  catalog_entry.unit_locator,
+                                                  catalog_entry.url,
+                                                  catalog_entry.data)
+            except DoesNotExist:
+                logger.debug(_('Failed to find a catalog entry with relative path'
+                               ' "{rel}".'.format(rel=relative_path)))
+                request.setResponseCode(NOT_FOUND)
+            except MultipleObjectsReturned:
```

Comment:
We'll want to support matching on multiple catalog entries.  See previous comment.